File: solutions/003/Peter/main.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def preorderhelper(root, serial):
    if (root is None):
        # assuming None will not be an input val
        serial.append("None")
        return
    serial.append(root.val)
    preorderhelper(root.left, serial)
    preorderhelper(root.right, serial)


def preorder(root):
    seriallist = list()
    preorderhelper(root, seriallist)
    returnstring = ""
    # print(seriallist)
    for node in seriallist:
        returnstring += node
        returnstring += " "
    return returnstring


def serialize(root):
    return preorder(root)


def deserializehelper(seriallist):
    cur = seriallist[0]
    seriallist.pop(0)
    if (cur == "None"):
        return None
    curNode = Node(cur)
    curNode.left = deserializehelper(seriallist)
    curNode.right = deserializehelper(seriallist)
    return curNode


def deserialize(serialstring):
    seriallist = serialstring.split(" ")
    return deserializehelper(seriallist)


node = Node('root', Node('left', Node('left.left')), Node('right'))
# print(serialize(node))
assert deserialize(serialize(node)).left.left.val == 'left.left'
assert deserialize(serialize(node)).right.val == 'right'

File: solutions/003/Peter/main.cpp
#include <iostream>
#include <sstream>
using namespace std;

class Node{
    public:
        string val;
        Node *left;
        Node *right;

        Node(string v, Node *l, Node *r){
            val = v;
            left = l;
            right = r;
        }
};

void serializeHelper(Node *n, stringstream &ss){
    if (n == nullptr) {
        // assuming nullptr will never be actual used as a value
        ss << "nullptr ";
        return;
    }
    ss << n->val << " ";
    serializeHelper(n->left, ss);
    serializeHelper(n->right, ss);
}

string serialize(Node *root){
    stringstream ss;
    serializeHelper(root, ss);
    return ss.str();
}

Node *deserializeHelper(stringstream &ss){
    string cur;
    ss >> cur;
    if (cur == "nullptr") return nullptr;
    Node* curNode = new Node(cur, nullptr, nullptr);
    curNode->left = deserializeHelper(ss);
    curNode->right = deserializeHelper(ss);
    return curNode;
}

Node *deserialize(string s){
    stringstream ss(s);
    Node *root = deserializeHelper(ss);
    return root;
}

//unneeded
Node *createNode(string v, Node *l, Node *r){
    Node *newN = new Node(v, l, r);
    return newN;
}

int main(){
    Node *root = new Node("root", new Node("left", new Node("left.left", nullptr, nullptr), nullptr), new Node("right", nullptr, nullptr));
    //cout << root->val;
    //cout << serialize(root) << endl;
    Node *root2 = deserialize(serialize(root));
    //cout << serialize(root2) << endl;
    assert(root2->left->left->val == "left.left");
    cout << "assert passed" << endl;
}